docs(animations): add section about animating reordering list items

dario-piotrowicz opened the pull request:
add a section regarding reordering list items in the complex animation
sequences guide to help developers rememeber to use a `TrackByFunction`
whenever they are animating `*ngFor` list items which change their
ordering

as suggested here: https://github.com/angular/angular/issues/42750#issuecomment-979127165

relates to issue #28040 and #42750

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## Issue

Issue Number: N/A


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No



## Other information
 - As suggested by @petebacondarwin 😄 
 - I _think_ the complex animation sequences guide is the right place for this section, please let me know if you disagree 🙂  
 - @jessicajaniuk 🙂

dario-piotrowicz committed:
docs(animations): add section about animating reordering list items

add a section regarding reordering list items in the complex animation
sequences guide to help developers rememeber to use a `TrackByFunction`
whenever they are animating `*ngFor` list items which change their
ordering

as suggested here: https://github.com/angular/angular/issues/42750#issuecomment-979127165

relates to issue #28040 and #42750

dario-piotrowicz commented:
I'm pretty sure that this also relates to: https://github.com/angular/angular/issues/19855 and https://github.com/angular/angular/issues/18847 (those are old issues with non-working reproductions in plunker)

crisbeto commented:
Shouldn't we try to fix the issue instead of adding notes about it in the docs?

dario-piotrowicz commented:
> Shouldn't we try to fix the issue instead of adding notes about it in the docs?

Thanks a bunch for the comment 🙂 

I think that the `trackBy` is actually necessary and there is nothing we can do about it, as the animation is based on the DOM and for list reordering Angular removes and re-inserts DOM elements so if we do not provide extra information for tracking the elements nothing can be done to know which element is which (based on their position alone)

I think this has already been discussed, like here: https://github.com/angular/angular/issues/40461#issuecomment-763240508 and here: https://github.com/angular/angular/issues/40461#issuecomment-763445767 and in the whole issue as a whole, from resolving the issue @IgorMinar added some details about this in the [`TrackByFunction` Api Docs](https://angular.io/api/core/TrackByFunction#description) (in PR https://github.com/angular/angular/pull/42329)

I am just adding the same sort of information in the guides as people do seem not to notice this in the `trackByFunction` api docs and end up confused as to why their list animations do no work properly

crisbeto commented:
My main concern is that we end saying that "it's just the way it works" and the bugs stay open forever.

Also I'm not completely sure that the issue based purely on the fact that we store information in the DOM. I have [this repro Stackblitz](https://stackblitz.com/edit/angular-ivy-24ym2t) that breaks when you reorder the array and the element has a child with an `ngIf`, but it starts working when you comment out the `ngIf`. I spent some time investigating it today and it looks like our logic for detecting when an element has been moved has issues.

dario-piotrowicz commented:
> My main concern is that we end saying that "it's just the way it works" and the bugs stay open forever.
> 
> Also I'm not completely sure that the issue based purely on the fact that we store information in the DOM. I have [this repro Stackblitz](https://stackblitz.com/edit/angular-ivy-24ym2t) that breaks when you reorder the array and the element has a child with an `ngIf`, but it starts working when you comment out the `ngIf`. I spent some time investigating it today and it looks like our logic for detecting when an element has been moved has issues.

Yes I totally agree...

The elements move operations are not optimal in the least, as for example I also have a PR open for an issue related to that: https://github.com/angular/angular/pull/44578

There is also a comment in the animations package saying that move animations are not supported: https://github.com/angular/angular/blob/38ddae1c2e0de87d5672fda30d4fbe567dab4123/packages/animations/browser/src/render/transition_animation_engine.ts#L994

About the `*ngFor` I just assumed that it was clear and understood that this is just how it has to be given Angular's implementation...

And I (personally, which is not saying much :P) cannot see any way to actually have this working without a substantial refactoring of how elements are reorders/moved...

Anyways what do you think then? what would you like to do? 🙂  should I close this PR (or convert it to draft) and do some investigation on how/if things can potentially be changed?

aikidave commented:
@TMDavisGoogle Can you please do an editing pass on this update? Thanks!

TMDavisGoogle commented on aio/content/guide/complex-animation-sequences.md:
```diff
@@ -88,6 +88,16 @@ For each change:
 
+## Animating the items of a reordering list
+
+Although Angular animates correctly `*ngFor` list items out of the box, it will not be able to do so if their ordering changes, this is because it will lose track of which element is which, resulting in broken animations. The only way to help Angular keep track of such elements is by assigning a `TrackByFunction` to the `NgForOf` directive, this makes sure that Angular always knows which element  ...
```
if their ordering changes, this is because it will 
```suggestion
Break up into separate sentences: "...if their ordering changes. This is because it will..." 
```

TMDavisGoogle commented on aio/content/guide/complex-animation-sequences.md:
```diff
@@ -88,6 +88,16 @@ For each change:
+
+<div class="alert is-important">
+
+**Rule of Thumb:** If you need to animate the items of an `*ngFor` list and there is a possibility that the order of such items will change during runtime always use a `TrackByFunction`.
```
the order of such items will change during runtime always use a `TrackByFunction`.
```suggestion
Insert comma after "runtime": "...the order of such items will change during runtime, always use a `TrackByFunction`."
```

TMDavisGoogle commented on aio/content/guide/complex-animation-sequences.md:
```diff
@@ -88,6 +88,16 @@ For each change:
+
 ## Animation sequence summary
 
 Angular functions for animating multiple elements start with `query()` to find inner elements, for example gathering all images within a `<div>`. The remaining functions, `stagger()`, <code>[group](api/animations/group)()</code>, and `sequence()`, apply cascades or lets you control how multiple animation steps are applied.
```
to find inner elements, for example gathering
```suggestion
Replace comma with semicolon after "elements" and add comma after "for example": "...to find inner elements; for example, gathering all..." 
```

TMDavisGoogle reviewed (commented):
Light edits (punctuation, sentence flow).

dario-piotrowicz committed:
fixup! docs(animations): add section about animating reordering list items

dario-piotrowicz committed:
docs: improve punctuation in animation sequence summary

slightly improve the punctuation in the animation sequence summary present
in the complex animation sequences guide

dario-piotrowicz commented:
@aikidave and @TMDavisGoogle thanks you very much for your help :slightly_smiling_face: 

I've applied the suggestion proposed by Tiffany although I'm not sure if this documentation update will land for the reasons highlighted by @crisbeto (but thank you so very much for looking at this regardless :slightly_smiling_face:)

aikithoughts reviewed (approved):
Reviewed-for: global-docs-approvers

All good from a writing perspective.

jessicajaniuk reviewed (approved):
LGTM :cookie:

IMHO, while I agree with @crisbeto, whether we fix the issue or not, having clearer documentation in the meantime about how / why it behaves this way is still better than the alternative. I feel we should merge this and then look at actually fixing the actual behavior.

dario-piotrowicz commented:
> LGTM cookie
> 
> IMHO, while I agree with @crisbeto, whether we fix the issue or not, having clearer documentation in the meantime about how / why it behaves this way is still better than the alternative. I feel we should merge this and then look at actually fixing the actual behavior.

Thanks a bunch :smile: 

Yeah I also agree with @crisbeto so whatever we decide to do with this PR I'll be happy with :slightly_smiling_face::+1:

dylhunn commented:
This PR was merged into the repository by commit f7ea275dc47d2fc1efbb3a83fe433c75ac3ed837.
